# 4.5 Data Consistency Checks

### This script contains:

#### 1. Import libraries
#### 2. Import dataframes
#### 3. Test dataframe
#### 4. Finding missing values
#### 5. Fixing missing values
#### 6. Duplicates
#### 7. Tasks

# Import libraries

In [1]:
import pandas as pd
import numpy as np
import os

# Import dataframes

In [2]:
# project folder path
path = r'C:\Users\Odette\Desktop\CareerFoundry\Immersion Courses\Course 4\Instacart Basket Analysis'

In [3]:
# import of products data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [5]:
# import of orders data
df_ords = pd.read_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [6]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_last_order  
0                  8                    NaN  
1                  7                   15.0  
2                 12                   21.0  
3                  7                   29.0  
4                 15                   28.0

# Test dataframe

In [7]:
# Create a dataframe
df_test = pd.DataFrame()

In [8]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [10]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [11]:
# Changing data type to string
df_test['mix'] = df_test['mix'].astype('str')

# Finding missing values

In [12]:
# Checking for missing vales is df_prods
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
# Creating a missing values subset
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# Fixing missing values

In [19]:
# How many rows in current df_prods dataframe
df_prods.shape

(49693, 5)

In [20]:
# Creating new dataframe with non-missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [23]:
# How many rows in new dataframe
df_prods_clean.shape

(49677, 5)

# Duplicates

In [26]:
# Finding full duplicates in the df_prods_clead dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [25]:
# Duplicates subset
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [27]:
# Current rows in df_prods_clean
df_prods_clean.shape

(49677, 5)

In [28]:
# New prods dataframe with no duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [30]:
df_prods_clean_no_dups.shape

(49672, 5)

# Tasks

In [31]:
# 2. describing df_prods_clean_no_dups
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

The max prices field looks too high at 99,999 and should be looked into

In [32]:
# 3. check for mixed data types in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

#4. The df_ords dataframe does not have any mixed data columns

In [33]:
# 5. rows in current df_ords dataframe
df_ords.shape

(3421083, 7)

In [34]:
# checking for missing vales is df_ords
df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

In [35]:
# creating a missing values subset
df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [36]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 7 columns]

There are 206,209 missing values. As these are all in the days_since_last_order column, I think these indicate the very first order of every user_id. If it is their first order, then there haven't been any days since their last order

#6. We cannot remove/filter the missing data as it is not actually missing, but indicates the very first order of every user_id.  There is no need to flag the missing values either. Instead, we need to impute the missing values, and replace them with '0'

In [46]:
# defining a new dataframe: df_ords_clean
df_ords_clean = df_ords

In [ ]:
# creating new dataframe 
df_ords_clean['days_since_last_order'].fillna(0, inplace=True)

In [ ]:
df_ords_clean.head(15)

In [50]:
# 7. finding full duplicates in the df_ords_clean dataframe
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [51]:
# duplicates subset
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order, first_order]
Index: []

There are no full duplicates in the df_ords_clean dataframe

#8. There are no duplicates to address

In [53]:
# 9. exporting df_prods_clean_no_dups dataframe as “products_checked.csv”
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'), index = False)

In [54]:
# 9. exporting df_ords_clean dataframe as “orders_checked.csv”
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'), index = False)